Modify CronJob API to add job history limits, cleanup jobs in controller

peay opened the pull request:
**What this PR does / why we need it**:
As discussed in #34710: this adds two limits to `CronJobSpec`, to limit the number of finished jobs created by a CronJob to keep.

**Which issue this PR fixes**: fixes #34710

**Special notes for your reviewer**:

cc @soltysh, please have a look and let me know what you think -- I'll then add end to end testing and update the doc in a separate commit. What is the timeline to get this into 1.6?

The plan:

- [x] API changes
  - [x] Changing versioned APIs
    - [x] `types.go`
    - [x] `defaults.go` (nothing to do)
    - [x] `conversion.go` (nothing to do?)
    - [x] `conversion_test.go` (nothing to do?)
  - [x] Changing the internal structure
    - [x] `types.go`
    - [x] `validation.go`
    - [x] `validation_test.go`
  - [x] Edit version conversions
    - [x] Edit (nothing to do?)
    - [x] Run `hack/update-codegen.sh`
  - [x] Generate protobuf objects
    - [x] Run `hack/update-generated-protobuf.sh`
  - [x] Generate json (un)marshaling code
    - [x] Run `hack/update-codecgen.sh`
  - [x] Update fuzzer
- [x] Actual logic
- [x] Unit tests
- [x] End to end tests
- [x] Documentation changes and API specs update in separate commit


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Add configurable limits to CronJob resource to specify how many successful and failed jobs are preserved.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40932)
<!-- Reviewable:end -->

peay commented:
Hmm, strange -- I rebased right before pushing this PR and there are still conflicts.

soltysh commented:
@peay thanks for the PR, I'll look into it first thing in the morning tomorrow :)

soltysh commented on pkg/controller/cronjob/cronjob_controller_test.go:
```diff
@@ -320,6 +330,188 @@ func TestSyncOne_RunOrNot(t *testing.T) {
+		sj.Spec.SuccessfulJobsHistoryLimit = &tc.successfulJobsHistoryLimit
+		sj.Spec.FailedJobsHistoryLimit = &tc.failedJobsHistoryLimit
+
+		// no deadline
```
I don't think it's needed.

peay replied:
Removed.

soltysh commented on pkg/controller/cronjob/utils.go:
```diff
@@ -234,6 +234,15 @@ func makeCreatedByRefJson(object runtime.Object) (string, error) {
+		}
+	}
+	return false, ""
+}
```
Can you combine `getFinishedStatus` with `IsJobFinished` in the following way:
```go
func IsJobFinished(j *batch.Job) bool {
	isFinished, _ = GetFinishedStatus(j)
	return isFinished
}
```
This will remove the code duplication. Btw. make the `GetFinishedStatus` public.

peay replied:
Done.

soltysh reviewed (changes requested):
Other than what I've mentioned above please keep the actual changes you did in one commit and the generated ones in a separate for ease of review :)

peay commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,10 +120,92 @@ func (jm *CronJobController) SyncAll() {
 	for _, sj := range sjs {
-		SyncOne(sj, jobsBySj[sj.UID], time.Now(), jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
+		jobsRemaining := CleanupFinishedJobs(sj, jobsBySj[sj.UID], jm.jobControl, jm.recorder)
+		SyncOne(sj, jobsRemaining, time.Now(), jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
```
OK, that's a good point -- this'll also eliminate the logic to maintain the list of jobs remaining, I actually like it better

peay replied:
In principle, I am not sure I can avoid to remove from active list though -- if the history limit is set to 0 (or to a low value, or there are many cronjobs maybe), cleanup may remove a job that finished in between syncing the cronjob and cleaning it up, which would still lead to an undesirable warning next time the cronjob is sync-ed.

Since I need to reap the pods manually, how about taking the controller's reaping code after
```.go
	if sj.Spec.ConcurrencyPolicy == batch.ReplaceConcurrent {
		for _, j := range sj.Status.Active {
```
out in a utility function, and also using that? This removes the pods, the job, and the reference in active list.

soltysh replied:
> In principle, I am not sure I can avoid to remove from active list though -- if the history limit is set to 0 (or to a low value, or there are many cronjobs maybe), cleanup may remove a job that finished in between syncing the cronjob and cleaning it up, which would still lead to an undesirable warning next time the cronjob is sync-ed.

Hmmm... either that or ignoring jobs that are active, but that on the other hand might complicate the removal logic. I'm fine with either.

> Since I need to reap the pods manually, how about taking the controller's reaping code after

Go for it!

peay commented:
Thanks for the review, I'll try to make the changes today!

peay commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,10 +120,92 @@ func (jm *CronJobController) SyncAll() {
+		j := js[i]
+		glog.Infof("Removing job %s from %s", j.Name, nameForLog)
+
+		// Do not remove pods, this will be done by GC as they
```
OK, nevermind, I just realized that "orphaned" in the GC controller means without a node, not without a parent.

smarterclayton commented on pkg/apis/batch/types.go:
```diff
@@ -244,6 +244,16 @@ type CronJobSpec struct {
+	// The number of failed finished jobs to retain.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// +optional
+	FailedJobsHistoryLimit *int32
```
Is retaining successful or failed jobs more likely in practice?  Or are there arguments for either setup?

peay replied:
I think for CronJob, which can generate a lot of successful jobs, it makes more sense to use two separate limits for success/failure.

If we use a single limit, with CronJobs that have a frequent schedule and in the normal situation where failures are infrequent, failed jobs are likely to get cleaned up before someone has time to look at them as more successful jobs pile up every minute or 5min or so.

soltysh replied:
It was also mentioned [here](https://github.com/kubernetes/kubernetes/issues/41039#issuecomment-277963383) that it's reasonable to have some defaults in place. @smarterclayton do you think 3 for successful and 1 for failed is a reasonable default?

smarterclayton replied:
Those numbers do not seem unreasonable.

soltysh replied:
@peay mind applying above defaults, in that case?

peay replied:
OK, will do

peay commented:
@soltysh updated the PR with:
- `Cleanup` after `Sync`
- Use reaping code from Replace concurrency policy also for cleanup
- Added a minor test variation
- Added Swagger/OpenAPI re-generation diff
- Misc. cosmetics and logging modifcations that you suggested

peay commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,10 +120,92 @@ func (jm *CronJobController) SyncAll() {
+	}
+
+	finishedFailedJobs := []batch.Job{}
+	finishedSuccessfulJobs := []batch.Job{}
```
Changed.

peay commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,10 +120,92 @@ func (jm *CronJobController) SyncAll() {
+func RemoveJobs(sj batch.CronJob, js []batch.Job, jc jobControlInterface, maxJobs int32, recorder record.EventRecorder) []batch.Job {
+	nameForLog := fmt.Sprintf("%s/%s", sj.Namespace, sj.Name)
+	numToDelete := len(js) - int(maxJobs)
+	glog.Infof("Cleaning up %d/%d jobs from %s", numToDelete, len(js), nameForLog, numToDelete)
```
Changed.

peay commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,10 +120,92 @@ func (jm *CronJobController) SyncAll() {
+	sort.Sort(byJobStartTime(js))
+	for i := 0; i < numToDelete; i++ {
+		j := js[i]
+		glog.Infof("Removing job %s from %s", j.Name, nameForLog)
```
Changed.

peay commented:
@soltysh updated the PR with:
- default values
- end to end testing

I will send a PR to `kubernetes.github.io` for documentation update.

kargakis commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,15 +120,73 @@ func (jm *CronJobController) SyncAll() {
+}
+
+// CleanupFinishedJobs cleanups finished jobs created by a CronJob
+func CleanupFinishedJobs(sj batch.CronJob, js []batch.Job, jc jobControlInterface, sjc sjControlInterface, pc podControlInterface, recorder record.EventRecorder) {
```
Unexport all functions that are not used outside of the package

peay replied:
Fixed, thanks!

peay commented:
@soltysh @kargakis updated the PR with:

- do not export `deleteJob`, `cleanupFinishedJobs`, `removeOldestJobs`
- un-exported pre-existing functions `syncOne`, `syncAll` for consistency

soltysh commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,15 +120,73 @@ func (jm *CronJobController) SyncAll() {
 	for _, sj := range sjs {
-		SyncOne(sj, jobsBySj[sj.UID], time.Now(), jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
+		sj := syncOne(sj, jobsBySj[sj.UID], time.Now(), jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
+		cleanupFinishedJobs(sj, jobsBySj[sj.UID], jm.jobControl, jm.sjControl, jm.podControl, jm.recorder)
```
Can both `syncOne` and `cleanupFinishedJobs` accept pointer to CronJob, that will get rid of the need to return anything from `syncOne`. I see we're not consistent with this along the code, eg. `removeOldestJob` does that, already.

soltysh commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,15 +120,73 @@ func (jm *CronJobController) SyncAll() {
+	succesfulJobs := []batch.Job{}
+
+	for _, job := range js {
+		isFinished, finishedStatus := GetFinishedStatus(&job)
```
`GetFinishedStatus` is private method used only by this controller, make it private.

soltysh commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,15 +120,73 @@ func (jm *CronJobController) SyncAll() {
+	sort.Sort(byJobStartTime(js))
+	for i := 0; i < numToDelete; i++ {
+		glog.V(4).Infof("Removing job %s from %s", js[i].Name, nameForLog)
+		_ = deleteJob(sj, &js[i], jc, pc, recorder, "history limit reached")
```
If you don't care about returned value, just don't assign it 😉

soltysh commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,15 +120,73 @@ func (jm *CronJobController) SyncAll() {
+	}
+
+	nameForLog := fmt.Sprintf("%s/%s", sj.Namespace, sj.Name)
+	glog.V(4).Infof("Cleaning up %d/%d jobs from %s", numToDelete, len(js), nameForLog, numToDelete)
```
Second `numToDelete`, at the end, is not necessary.

soltysh commented on pkg/controller/cronjob/cronjob_controller_test.go:
```diff
@@ -320,10 +330,198 @@ func TestSyncOne_RunOrNot(t *testing.T) {
+}
+
+func TestCleanupFinishedJobs_DeleteOrNot(t *testing.T) {
+	// Starting times are assumbed to be sorted by increasing start time
```
typo: assumbed/assumed

soltysh commented on test/e2e/cronjob.go:
```diff
@@ -225,10 +235,227 @@ var _ = framework.KubeDescribe("CronJob", func() {
 	})
+
+	// cleanup of successful finished jobs, with default limits
+	It("should delete successful finished jobs according to default history limits [Slow]", func() {
```
Appreciate the amount of e2e tests, but @spxtr is trying to decrease the time they run (see https://github.com/kubernetes/kubernetes/issues/40839 for discussion). Your new tests will significantly increase that number, unfortunately. I'd suggest one test with limits set to 1 and 1 and verify it's working as expected, the rest has already pretty good coverage in unit tests.

soltysh replied:
Make sure the test is fast, iow. instead of `sleep XXX` let it just print hello world, or alike.

soltysh replied:
You could also verify just one successful or failed, whichever you prefer, since the path for both goes through the same path.

soltysh reviewed (changes requested):
@peay more comments, but big thanks for working on this!

peay commented:
@soltysh updated the PR with:

- use pointers to CronJob in `SyncOne` to homogenise use in controllers -- utils still take values though
- un-export `GetFinishedStatus`
- fixed typos
- changed end to end testing:
  - only two tests: one for failed, one for successful (I don't see how to do both in one test only, but these did run in 120-150s during my tests)
  - make these two tests faster using `/bin/true` and `/bin/false` commands instead of `sleep`

soltysh commented on test/e2e/cronjob.go:
```diff
@@ -225,10 +239,92 @@ var _ = framework.KubeDescribe("CronJob", func() {
+		jobs, err := f.ClientSet.Batch().Jobs(f.Namespace.Name).List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		_, finishedJobs := filterActiveJobs(jobs)
+		Expect(len(finishedJobs) >= 1).To(BeTrue())
```
There should not be anymore than just one job at any point in time, almost. The only time is not true is right after a new got scheduled and an old wasn't cleaned, yet. But I doubt you hit that point. Check for one and expect that one being removed in the next step.

soltysh commented on test/e2e/cronjob.go:
```diff
@@ -225,10 +239,92 @@ var _ = framework.KubeDescribe("CronJob", func() {
+	})
+
+	// cleanup of failed finished jobs, with limit of one failed job
+	It("should delete failed finished jobs with limit of one failed job", func() {
```
Drop it entirely, just single e2e test is sufficient here, especially that the code is copy&paste.

soltysh commented on test/e2e/cronjob.go:
```diff
@@ -319,6 +424,28 @@ func waitForNoJobs(c clientset.Interface, ns, jobName string, failIfNonEmpty boo
 }
 
+// Wait for a list of jobs to not exist by listing jobs explicitly.
+func waitForJobsNotExist(c clientset.Interface, ns string, jobs []*batchv1.Job) error {
```
Suggestion: `waitForJobNotExist` we care about one, as stated before. That will also simplify the code, b/c you'll get single job by name.

soltysh commented on test/e2e/cronjob.go:
```diff
@@ -271,8 +367,17 @@ func newTestCronJob(name, schedule string, concurrencyPolicy batch.ConcurrencyPo
+	if failedJobsHistoryLimit != nil {
+		sj.Spec.FailedJobsHistoryLimit = failedJobsHistoryLimit
+	}
+	if activeDeadlineSeconds != nil {
```
Drop it, not needed if you skip the second test.

soltysh commented on test/e2e/cronjob.go:
```diff
@@ -271,8 +367,17 @@ func newTestCronJob(name, schedule string, concurrencyPolicy batch.ConcurrencyPo
 	}
-	if sleep {
-		sj.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Command = []string{"sleep", "300"}
+	if successfulJobsHistoryLimit != nil {
```
These are both pointers, so you can assign it always, they should get defaulted properly by the api server.

soltysh reviewed (changes requested):
One last round of comments and I think we're there. By the way, I forgot to mention that last time, please squash all your generated changes into a single commit.

peay commented:
@soltysh Thanks for the comments, all good points. Updated the PR with:

- Only test successful job case
- Simplify logic with `waitForJobsNotExist` instead of list of jobs
- Removed `failedJobsHistoryLimit` or `activeDeadlineSeconds` in `newTestCronJob`
- Single commit for generated changes (However, should https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#examples-and-docs be updated? This mentions  API spec update should be in a separate commit if I understand it right)

soltysh commented:
@k8s-bot ok to test

soltysh commented:
/lgtm

@peay huge thanks!!!

soltysh commented:
@smarterclayton can this get your approval?

peay commented:
Hum, I had carefully tested everything again and again locally, but it looks like there are some issue with Bazel, with OpenAPI specs, and with the round trip tests. Not quite sure how that is.

`make verify` passes on my machine, and says everything is up to date: bazel, swagger/openapi specs, and generated code. Manually running `update-bazel.sh` or the openapi scripts does not change any file.

`make test WHAT=k8s.io/kubernetes/pkg/api KUBE_GOFLAGS="-v"` also passes on my machine, and  correctly logs `serialization_test.go:236: working on CronJobList in batch` for `TestRoundTripTypes`. 

@k8s-bot bazel test this

soltysh commented:
/lgtm

peay commented:
@soltysh when do you think we can merge this? I'll rebase it again but just want to avoid doing it too many times. Thanks!

soltysh commented:
@peay this needs approval from Clayton (or someone from api-reviewers), if you rebase it, I'll try to talk to him later today and get this approval

peay commented:
OK, running tests and verify right now, I'll push in about half an hour.

soltysh commented:
I'll wait for green tests and will apply lgtm.

soltysh commented:
/lgtm

smarterclayton commented on pkg/apis/batch/v2alpha1/types.go:
```diff
@@ -250,6 +250,16 @@ type CronJobSpec struct {
+	// The number of successful finished jobs to retain.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// +optional
+	SuccessfulJobsHistoryLimit *int32 `json:"successfulJobsHistoryLimit,omitempty" protobuf:"varint,6,opt,name=successfulJobsHistoryLimit"`
```
Document the defaults for both of these fields in the godoc.

Also, what happens when upgrading a cluster?  What if someone has failed jobs that they want to keep?  Will we suddenly purge them?

soltysh replied:
> Document the defaults for both of these fields in the godoc.

I haven't seen we document those defaults in the api, elsewhere. 



> Also, what happens when upgrading a cluster? What if someone has failed jobs that they want to keep? Will we suddenly purge them?

It looks like we will, but that's the whole point of introducing these with some reasonable values. We can't keep everything forever.

smarterclayton replied:
The vast majority of defaults are documented.  How else are users supposed to know what they are? 

Since that's potentially a data loss issue, you'll need to make very certain that the impacts of this are understood.  @kargakis did we do any prep for deployment cleanup?  At minimum the important release notes.

soltysh replied:
The other options is, since we're introducing `batch/v1beta1.CronJob` we could introduce these defaults there, and let the `batch/v2alpha1.CronJob` just introduce these fields, but without defaulting, wdyt?

kargakis replied:
@smarterclayton @soltysh is adding the defaults only for Deployments created in new clusters. Older Deployments shouldn't be pruned. https://github.com/kubernetes/kubernetes/pull/39683

soltysh replied:
Yeah, that was my proposal as well. We can introduce these fields, but only apply them to the new CronJobs created through new endpoint. I'm guessing Clayton's silence can be seen as a approval for this type of approach.

smarterclayton replied:
Yes.

soltysh commented:
@peay sorry to bother you again, I've just removed the lgtm label, can you please remove the defaulting part from this PR, I'll pick it up in the next PR that comes after this one, which is promoting CronJobs to beta to satisfy the concerns @smarterclayton raised.

peay commented:
@soltysh yes, sure, will do this within a day or two, currently traveling

soltysh commented:
Ok, big thanks!

peay commented:
@soltysh updated PR with:
- removed defaults in `SetDefaults_CronJob`
- modified fuzzer to randomize whether each limit is set

`podautoscaler` failures and the like seem to be flakes.

@k8s-bot test this

peay commented:
@k8s-bot e2e test this

soltysh commented:
@k8s-bot non-cri e2e test this

soltysh commented:
/lgtm

Thanks @peay!

@smarterclayton are we good for approval, the defaults (as discussed above) will be applied to the new endpoint, I'm currently working on.

soltysh commented:
@k8s-bot cvm gce e2e test this

soltysh commented:
@k8s-bot gci gce e2e test this

soltysh commented:
@k8s-bot gce etcd3 e2e test this

soltysh commented:
@k8s-bot non-cri e2e test this

soltysh commented:
@k8s-bot gci gce e2e test this

peay commented:
@soltysh still traveling and haven't had time to really check but it seems like the failures are still not related to this PR, right?

soltysh commented:
@peay, that's what I'm thinking as well, mind rebasing the PR, in the mean time. I remember there were some problems with tests, recently. I'll dig into this more when it fails after the rebase.

peay commented:
@soltysh yes, done

peay commented:
@k8s-bot bazel test this

soltysh commented:
@k8s-bot bazel test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot cvm gce e2e test this

soltysh commented:
/lgtm

smarterclayton commented:
API change approved

/approve

soltysh commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this

soltysh commented:
@k8s-bot non-cri e2e test this

soltysh commented:
@k8s-bot non-cri e2e test this

sttts commented:
@k8s-bot gci gce e2e test this

soltysh commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -119,24 +120,82 @@ func (jm *CronJobController) SyncAll() {
+		failedJobs,
+		jc,
+		pc,
+		*sj.Spec.FailedJobsHistoryLimit,
```
This is panic-ing e2e tests, because it's not set. Add condition around each `removeOldestJobs` checking if it's set and only then remove jobs. Make sure to change your unit test to verify when only one is set, currently you're accepting `int32` for both without being able to set nil, make sure to change that and test that condition.

soltysh commented:
@peay I found the root cause, thanks to @sttts ❤️ . Please update this PR asap or if you can't lemme know I'll do it.

peay committed:
Modify CronJob API to add job history limits, cleanup jobs in controller

peay committed:
Re-generate code and API spec for CronJob API

peay commented:
@soltysh good catc, modified the PR with:
- check whether limit is set before attempting to delete
- added unit test cases for when either or both limits are unset (these are two different paths as we shortcut to avoid listing jobs when both are unset)

soltysh commented:
/lgtm

peay commented:
@k8s-bot bazel test this

sttts commented:
@k8s-bot gce etcd3 e2e test this
